DM-7269: Port to Python 3

r-owen opened the pull request.

r-owen committed:
Run autopep8 on all Python code

r-owen committed:
Run futurize -1 -w .

r-owen committed:
Add _build.* and tests/.cache to .gitignore

r-owen committed:
Run futurize -2 -x division_safe -w .

r-owen committed:
Add python_future as a requirement to the ups table file

r-owen committed:
Minor clean ups in tests

Have tearDown delete LSST objects created in setUp
Break lines that are too long
Standardize lists of expected strings
Remove unused imports

r-owen committed:
PipelineProvenanceRecorder pep8 fixes

Break long lines

r-owen committed:
Modernize how tests are run for pytest

timj commented on python/lsst/ctrl/provenance/dc3.py:
```diff
@@ -155,7 +159,7 @@ def queryRunOffset(self):
         self._globalDb.setQueryWhere("runId = :runId")
         self._globalDb.query()
-        if not self._globalDb.next() or self._globalDb.columnIsNull(0):
+        if not next(self._globalDb) or self._globalDb.columnIsNull(0):
```
This might be able to throw a `StopIteration` exception but I'm not entirely sure what `daf_persistence` does with `DbStorage` when it runs out of items.

timj replied:
I don't see any tests where we test `next` as the specific tests are disabled in `daf_persistence`. @n8pease it looks like my fix for DM-7226 might be wrong. I am thinking that we need to trap for the final item and then raise `StopIteration` if we want to be compliant with the Python expectation.

timj replied:
PS @srp3rd I tried running the `daf_persistence` tests on `lsst-dev` and they can't connect to the server either so I'm guessing the server is down?

timj commented on tests/testProvenanceSetup.py:
```diff
@@ -132,15 +136,15 @@ def testAddCmd(self):
-        
+
         self.setup.addWorkflowRecordCmd("dbingest.py", "-u ray".split(),
                                         "/usr/local/bin/dbingest.py")
```
we install files into `/usr/local/`?

r-owen replied:
@srp3rd do you have an opinion about this? It looks suspicious to me (though if it is to be fixed, let's use a new ticket)

srp3rd replied:
Is that really what it's doing?  It appears to me that it's just testing adding a command, not actually executing anything there.

r-owen replied:
I think the question is "is it reasonable to add a command to /usr/local/bin"? I would expect commands in ctrl_provenance to add commands to `os.path.join(getPackageDir("ctrl_provenance"), "bin")` if the command needs any software managed by eups. (It also seems odd to add a command and not verify that it was added, or at least that the file exists, but that's another story).

srp3rd replied:
The code itself only manipulates lists and commands, and never tests were anything is.  It could be "/usr/foobar/woka/woka/dbingest.py" and it would work.   In practice, the code you gave above would be the correct way to invoke the API.  I don't see the harm in having the test the way it is.

I believe that the reason the code exists in this way is that provenance (at least when it was used in orchestration), records where the executables should be on the remote resource.  The ctrl_provenance code didn't execute on the remote resource, only from the launch platform.   Is this ideal?  I don't think it is, but it's what was done at the time.

timj replied:
Ok. Seems like a none issue.

timj replied:
Can we add a comment to this code though?

r-owen replied:
A comment sounds wise. Unfortunately I don't understand Steve's explanation well enough to document what is going on. (Is the code is actually putting a file in /usr/local/bin or just telling provenance to record that a file exists in /usr/local/bin?). I'm not complaining about the code, just saying I'm not too ignorant to explain it.

srp3rd replied:
It's telling provenance to record that the path full path to the executable is at that location.  Provenance doesn't verify that it is there.   I added a comment.

timj commented:
Looks like the python3 part of this is trivial. I have concerns over `StopIteration` and that our `daf_persistence.DbStorage` implementation may not be quite right.

r-owen committed:
Fix unwanted change db.next() to next(db)

DbStorage has a next method that is not a python iterator,
so call using db.next, not next(db).

r-owen committed:
Fix flake8 warnings

Import missing modules.
Remove unused imports.
Remove trailing whitespace.
Stop one instance of assigning to an unused variable.

r-owen committed:
Standardize order and grouping of imports

srp3rd committed:
Added further explanation to the addWorkflowRecordCmd method
about the "path" argument
